### PR TITLE
port apply fsm step 5

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

**Comment by coocood** [comment]
@zhangjinpeng1987 PTAL

File: tikv/raftstore/util.go [hunk 1]
```diff
@@ -456,3 +468,11 @@ func checkPeerID(req *raft_cmdpb.RaftCmdRequest, peerID uint64) error {
 	}
 	return errors.Errorf("mismatch peer id %d != %d", peer.Id, peerID)
 }
+
+func CloneMsg(origin, cloned proto.Message) error {
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/util.go]: I think cloned should be a pointer type?

Comment (coocood) [reply tikv/raftstore/util.go]: It's an interface, the interface contains a pointer.

Comment (zhangjinpeng1987) [reply tikv/raftstore/util.go]: got it

File: tikv/raftstore/fsm_apply.go [hunk 2]
```diff
@@ -1037,34 +1053,300 @@ func (d *applyDelegate) createWriteCmdOps(requests []*raft_cmdpb.Request) (ops w
...
+		if p := findPeer(region, storeID); p != nil {
+			exist = true
+			if !p.IsLearner || p.Id != peer.Id {
+				errMsg := fmt.Sprintf("%s can't add duplicated peer, peer %s, region %s",
+					d.tag, p, d.region)
+				log.Error(errMsg)
+				err = errors.New(errMsg)
+				return
+			} else {
+				p.IsLearner = true
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/fsm_apply.go]: Should be false.

File: tikv/raftstore/fsm_apply.go [hunk 3]
```diff
@@ -1037,34 +1053,300 @@ func (d *applyDelegate) createWriteCmdOps(requests []*raft_cmdpb.Request) (ops w
...
+	}
+	log.Infof("%s split region %s, keys %v", d.tag, d.region, keys)
+	derived.RegionEpoch.Version += uint64(newRegionCnt)
+	// Note that the split requests only contain ids for new regions, so we need
+	// to handle new regions and old region separately.
+	if !rightDerive {
+		derived.EndKey = keys[1]
+		keys = keys[1:]
+		regions = append(regions, derived)
+	}
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/fsm_apply.go]: Else you should drop the last key in keys?

Comment (coocood) [reply tikv/raftstore/fsm_apply.go]: No need, the last key will not be used.

**Pull request merged**